### PR TITLE
File Downloading and Report Changes

### LINKED ISSUE #698: Election Type in Intermediate Audit Report title shd match SoS Audit Def page
> State   Audit Report
> --
> Generated   2017-09-13T16:36:42.224
> coordinated - 2017-11-07

Instead of "coordinated" this should read "Coordinated Election" (etc., using strings from http://localhost:3000/sos/audit).

### LINKED ISSUE #749: List number of discrepancies in each round in County Audit Report
(per RLA Tool Feedback #1_tool version2)

The number of discrepancies should be listed by round [ in the County Audit Report ].

### LINKED ISSUE #751: Allow State user to download County-uploaded BM and CVR files
(Note: this is one component of #601 ).

Requested by CDOS in RLA Tool Feedback #1_tool version2.

### LINKED ISSUE #753: Remove audit_sequence_number from csv list of ballot cards to audit
CDOS requested in RLA Tool Feedback #1_tool version2.

### PR DESCRIPTION
**Pull request by dmzimmerman**
More features and fixes for today's milestone. In particular, file downloads (with the same basic interface as ballot manifest/cvr imports - the client supplies the file info, the endpoint returns it the file instead of importing the file), and report changes.

Closes #753
Closes #751 
Closes #749 
Closes #698

### CODE REVIEW

File: server/eclipse-project/src/main/java/us/freeandfair/corla/endpoint/FileDownload.java [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+   */
+  @Override
+  public String endpoint(final Request the_request, final Response the_response) {
+    // we know we have either state or county authentication; this will be null
+    // for state authentication
+    final County county = Main.authentication().authenticatedCounty(the_request);
+
+    try {
+      final UploadedFile file =
+          Main.GSON.fromJson(the_request.body(), UploadedFile.class);
```
[the remark this answers is not in the document]
Comment (ranweiler) [reply server/eclipse-project/src/main/java/us/freeandfair/corla/endpoint/FileDownload.java]: What I'd most like to see: an API where this endpoint takes two parameters, `countyId` and `fileType`, where `fileType` is either `cvrExport` or `ballotManifest`, and then returns the relevant file if it exists and the client is authorized to download it.

[the remark this answers is not in the document]
Comment (dmzimmerman) [reply server/eclipse-project/src/main/java/us/freeandfair/corla/endpoint/FileDownload.java]: I don't like this, and here's why: it gives no way of downloading old files that have been replaced. I do appreciate the idea that it should be a `GET`, but I think it needs to take all the parameters it currently takes.

[the remark this answers is not in the document]
Comment (ranweiler) [reply server/eclipse-project/src/main/java/us/freeandfair/corla/endpoint/FileDownload.java]: That makes sense, assume that is a feature we really do want. To push back a little: we don't expose _anything_ via the web app except for the _current_ file, if any. So, unless a user were tracking those data themselves out-of-band, I'm not sure there's actually a use case for that feature (I may be mistaken). WDYT?

[the remark this answers is not in the document]
Comment (dmzimmerman) [reply server/eclipse-project/src/main/java/us/freeandfair/corla/endpoint/FileDownload.java]: We don't expose anything like that now, but one can imagine (with not much imagination :) ) a useful endpoint that would list all the uploaded files for a county. If we define this download endpoint too restrictively, we'd have to end up redefining it later.

[the remark this answers is not in the document]
Comment (ranweiler) [reply server/eclipse-project/src/main/java/us/freeandfair/corla/endpoint/FileDownload.java]: Okay, how about: switch to a `GET` with query params, but the same params you are expecting in the JSON right now.

**Pull request merged**